File: tests/echo/echo.cpp
// Copyright (c) 2017 Timo Savola. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include <stdint.h>
#include <string.h>

#include <gate.h>

#include "../discover.h"

#define CONTENT "ECHO.. Echo.. echo..\n"

struct message_op {
	gate_packet header;
	char content[sizeof (CONTENT)];
} GATE_PACKED;

struct message_ev {
	gate_packet header;
	char content[0];
} GATE_PACKED;

static void send_message()
{
	const message_op packet = {
		.header = {
			.size = sizeof (packet),
		},
		.content = CONTENT,
	};

	gate_send_packet(&packet.header);
}

static message_ev *recv_message(char *buf, size_t bufsize)
{
	auto packet = reinterpret_cast<message_ev *> (buf);

	do {
		gate_recv_packet(buf, bufsize, 0);
	} while (packet->header.code != 0);

	return packet;
}

int main()
{
	discover_service("echo");
	send_message();

	char buf[gate_max_packet_size];
	const auto reply = recv_message(buf, gate_max_packet_size);

	auto length = reply->header.size - sizeof (reply->header);
	if (length != sizeof (CONTENT)) {
		gate_debug("Length mismatch\n");
		return 1;
	}

	if (memcmp(reply->content, CONTENT, length) != 0) {
		gate_debug("Content mismatch\n");
		return 1;
	}

	return 0;
}

File: service/echo/echo.go
// Copyright (c) 2017 Timo Savola. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package echo

import (
	"sync/atomic"

	"github.com/tsavola/gate/packet"
	"github.com/tsavola/gate/service"
)

const (
	Name    = "echo"
	Version = 0
)

var prevId uint64 // atomic

type Config struct {
	Log Logger
}

var Default = new(Config)

func (c *Config) Register(r *service.Registry) {
	r.Register(Name, Version, c)
}

func (c *Config) Instantiate(packet.Code, *service.Config) service.Instance {
	return &echo{
		Config: *c,
		id:     atomic.AddUint64(&prevId, 1),
	}
}

type echo struct {
	Config
	id uint64
}

func (e *echo) Handle(p packet.Buf, replies chan<- packet.Buf) {
	replies <- p

	if e.Log != nil {
		e.Log.Printf("instance %d: %#v", e.id, string(p.Content()))
	}
}

func (e *echo) Shutdown() {
	if e.Log != nil {
		e.Log.Printf("instance %d: shutdown", e.id)
	}
}
